### PR TITLE
GAPI Fluid: Dynamic window size

### PR DESCRIPTION
**Pull request by akhakim**
```
build_gapi_standalone:Linux x64=ade-0.1.1f
build_gapi_standalone:Win64=ade-0.1.1f
build_gapi_standalone:Mac=ade-0.1.1f
build_gapi_standalone:Linux x64 Debug=ade-0.1.1f
```

### CODE REVIEW

**Comment by akhakim** [comment]
@dmatveev @rgarnov

**Comment by akhakim** [comment]
build_gapi_standalone:Linux x64=ade-0.1.1f
build_gapi_standalone:Win64=ade-0.1.1f
build_gapi_standalone:Mac=ade-0.1.1f
build_gapi_standalone:Linux x64 Debug=ade-0.1.1f

**Comment by alalek** [comment]
> build_gapi_standalone

This can be applied through PR description message only (see your previous PR). Also there are no automatic triggering on properties change (only changed commits can trigger builds)

File: modules/gapi/src/backends/fluid/gfluidimgproc.cpp [hunk 1]
```diff
@@ -546,6 +556,18 @@ GAPI_FLUID_KERNEL(GFluidBoxFilter, cv::gapi::imgproc::GBoxFilter, true)
...
     }
+
+    static int getWindow( const cv::GMatDesc& /* src */,
+                          int       /* ddepth */,
+                          const cv::Size    & kernelSize,
+                          const cv::Point   & /* anchor */,
+                          bool      /* normalize */,
+                          int     /* borderType*/,
+                          const cv::Scalar  &  /*  borderValue*/)
+    {
```
[the remark this answers is not in the document]
Comment (akhakim) [reply modules/gapi/src/backends/fluid/gfluidimgproc.cpp]: Applied.

File: modules/gapi/src/backends/fluid/gfluidimgproc.cpp [hunk 2]
```diff
@@ -767,6 +789,19 @@ GAPI_FLUID_KERNEL(GFluidSepFilter, cv::gapi::imgproc::GSepFilter, true)
...
+                                int       /* ddepth */,
+                         const cv::Mat&       kernX ,
+                         const cv::Mat&       kernY ,
+                         const cv::Point&    /* anchor */,
+                         const cv::Scalar&   /* delta */,
+                            int         /* borderType*/,
+                         const cv::Scalar& /* borderValue*/)
+    {
+        GAPI_Assert((kernX.rows * kernX.cols) == (kernY.rows * kernY.cols));
+        return (kernX.rows * kernX.cols);
```
[the remark this answers is not in the document]
Comment (akhakim) [reply modules/gapi/src/backends/fluid/gfluidimgproc.cpp]: Applied.

File: modules/gapi/src/backends/fluid/gfluidbackend.cpp [hunk 3]
```diff
@@ -1506,6 +1506,8 @@ void GFluidBackendImpl::addBackendPasses(ade::ExecutionEngineSetupContext &ectx)
                 auto &fu = fg.metadata(node).get<FluidUnit>();
                 const auto &op = g.metadata(node).get<Op>();
 
+                fu.windowSize = fu.k.m_win(GModel::collectInputMeta(fg, node), op.args);
+
                 // Trigger user-defined "getBorder" callback
                 fu.border = fu.k.m_b(GModel::collectInputMeta(fg, node), op.args);
```
[the remark this answers is not in the document]
Comment (akhakim) [reply modules/gapi/src/backends/fluid/gfluidbackend.cpp]: Applied.

File: modules/gapi/include/opencv2/gapi/fluid/gfluidkernel.hpp [hunk 4]
```diff
@@ -304,6 +339,12 @@ struct FluidCallHelper<Impl, std::tuple<Ins...>, std::tuple<Outs...>, UseScratch
         constexpr bool callCustomGetBorder = (Impl::Window != 1);
         return get_border_helper<callCustomGetBorder, Impl, Ins...>::help(metas, in_args);
     }
+
+    static int getWindow(const GMetaArgs &metas, const cv::GArgs &in_args)
+    {
+        constexpr bool callCustomGetWindow = (Impl::Window > 1);
+        return get_window_helper<callCustomGetWindow, Impl, Ins...>::help(metas, in_args);
```
[the remark this answers is not in the document]
Comment (akhakim) [reply modules/gapi/include/opencv2/gapi/fluid/gfluidkernel.hpp]: Applied.

File: modules/gapi/include/opencv2/gapi/fluid/gfluidkernel.hpp [hunk 5]
```diff
@@ -304,6 +339,12 @@ struct FluidCallHelper<Impl, std::tuple<Ins...>, std::tuple<Outs...>, UseScratch
         constexpr bool callCustomGetBorder = (Impl::Window != 1);
         return get_border_helper<callCustomGetBorder, Impl, Ins...>::help(metas, in_args);
     }
+
+    static int getWindow(const GMetaArgs &metas, const cv::GArgs &in_args)
+    {
+        constexpr bool callCustomGetWindow = (Impl::Window > 1);
```
[the remark this answers is not in the document]
Comment (akhakim) [reply modules/gapi/include/opencv2/gapi/fluid/gfluidkernel.hpp]: Applied.

**Comment by akhakim** [comment]
@TolyaTalamanov

File: modules/gapi/src/backends/fluid/gfluidbackend.cpp [hunk 6]
```diff
@@ -178,6 +179,13 @@ struct FluidFilterAgent : public FluidAgent
     virtual void setRatio(double) override { /* nothing */ }
 public:
     using FluidAgent::FluidAgent;
+    int m_window;
+
+    FluidFilterAgent(const ade::Graph &g, ade::NodeHandle nh):FluidAgent(g, nh)
+    {
+        m_window = GConstFluidModel(g).metadata(nh).get<FluidUnit>().window;
+    }
+    virtual ~FluidFilterAgent() = default;
```
Comment (rgarnov) [inline modules/gapi/src/backends/fluid/gfluidbackend.cpp]: FluidAgent has a virtual destructor, this one should be removed

File: modules/gapi/src/backends/fluid/gfluidbackend.cpp [hunk 7]
```diff
@@ -178,6 +179,13 @@ struct FluidFilterAgent : public FluidAgent
     virtual void setRatio(double) override { /* nothing */ }
 public:
     using FluidAgent::FluidAgent;
+    int m_window;
+
+    FluidFilterAgent(const ade::Graph &g, ade::NodeHandle nh):FluidAgent(g, nh)
```
Comment (rgarnov) [inline modules/gapi/src/backends/fluid/gfluidbackend.cpp]: Please format `:` like `cv::gimpl::FluidAgent::FluidAgent`:
```
FluidFilterAgent(const ade::Graph &g, ade::NodeHandle nh)
    : FluidAgent(g, nh)
{
    m_window = GConstFluidModel(g).metadata(nh).get<FluidUnit>().window;
}
```

File: modules/gapi/src/backends/fluid/gfluidbackend.cpp [hunk 8]
```diff
@@ -34,6 +34,7 @@
 
 #include "backends/fluid/gfluidbuffer_priv.hpp"
 #include "backends/fluid/gfluidbackend.hpp"
+#include "../include/opencv2/gapi/fluid/gfluidkernel.hpp"
```
Comment (rgarnov) [inline modules/gapi/src/backends/fluid/gfluidbackend.cpp]: Is this include required?

File: modules/gapi/src/backends/fluid/gfluidbackend.cpp [hunk 9]
```diff
@@ -1505,9 +1513,13 @@ void GFluidBackendImpl::addBackendPasses(ade::ExecutionEngineSetupContext &ectx)
                 // FIXME: check that op has only one data node on input
                 auto &fu = fg.metadata(node).get<FluidUnit>();
                 const auto &op = g.metadata(node).get<Op>();
+                auto collectInputMeta = GModel::collectInputMeta(fg, node);
```
Comment (rgarnov) [inline modules/gapi/src/backends/fluid/gfluidbackend.cpp]: Why don't just name it `inputMeta`?

File: modules/gapi/src/backends/fluid/gfluidbackend.cpp [hunk 10]
```diff
@@ -1505,9 +1513,13 @@ void GFluidBackendImpl::addBackendPasses(ade::ExecutionEngineSetupContext &ectx)
...
+
+                // Trigger user-defined "getWindow" callback
+                fu.window = fu.k.m_gw(collectInputMeta, op.args);
 
                 // Trigger user-defined "getBorder" callback
-                fu.border = fu.k.m_b(GModel::collectInputMeta(fg, node), op.args);
+                fu.border = fu.k.m_b(collectInputMeta, op.args);
             }
```
Comment (rgarnov) [inline modules/gapi/src/backends/fluid/gfluidbackend.cpp]: Please rename the pass to `init_fluid_unit_windows_and_borders`

File: modules/gapi/src/backends/fluid/gfluidimgproc.cpp [hunk 11]
```diff
@@ -546,6 +556,18 @@ GAPI_FLUID_KERNEL(GFluidBoxFilter, cv::gapi::imgproc::GBoxFilter, true)
     {
         return { borderType, borderValue};
     }
+
+    static int getWindow(const cv::GMatDesc& /* src */,
```
Comment (rgarnov) [inline modules/gapi/src/backends/fluid/gfluidimgproc.cpp]: Are there some tests for added functionality? As I can see GFluidBlur still has static window defined so this callback will never be called

**Comment by rgarnov** [comment]
I see `compare ABI dumps` failure on BuildBot, that's expected

File: modules/gapi/src/backends/fluid/gfluidimgproc.cpp (hunk 11 above)
Comment (akhakim) [reply modules/gapi/src/backends/fluid/gfluidimgproc.cpp]: Applied.

File: modules/gapi/src/backends/fluid/gfluidbackend.cpp (hunk 10 above)
Comment (akhakim) [reply modules/gapi/src/backends/fluid/gfluidbackend.cpp]: Applied.

File: modules/gapi/src/backends/fluid/gfluidbackend.cpp (hunk 9 above)
Comment (akhakim) [reply modules/gapi/src/backends/fluid/gfluidbackend.cpp]: Applied.

File: modules/gapi/src/backends/fluid/gfluidbackend.cpp (hunk 8 above)
Comment (akhakim) [reply modules/gapi/src/backends/fluid/gfluidbackend.cpp]: Applied.

File: modules/gapi/src/backends/fluid/gfluidbackend.cpp (hunk 6 above)
Comment (akhakim) [reply modules/gapi/src/backends/fluid/gfluidbackend.cpp]: Applied.

File: modules/gapi/include/opencv2/gapi/fluid/gfluidkernel.hpp [hunk 12]
```diff
@@ -255,6 +258,67 @@ struct get_border_helper<false, Impl, Ins...>
...
+    static constexpr auto Check(U*) -> typename std::is_same<decltype(U::Window), T>::type;
+
+    template<typename>
+    static constexpr std::false_type Check(...);
+
+    typedef decltype(Check<C>(0)) Result;
+
+public:
+    static constexpr bool value = Result::value;
+};
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/fluid/gfluidkernel.hpp]: Cool

File: modules/gapi/src/backends/fluid/gfluidbackend.cpp [hunk 13]
```diff
@@ -312,11 +318,11 @@ static int maxLineConsumption(const cv::GFluidKernel& k, int inH, int outH, int
     }
 }
 
-static int borderSize(const cv::GFluidKernel& k)
+static int borderSize(const cv::GFluidKernel::Kind kind, int window)
```
Comment (dmatveev) [inline modules/gapi/src/backends/fluid/gfluidbackend.cpp]: So why you decided to go away from `cv::GFluidKenrel&` object here to more specific fields? Doesn't that object hold the necessary `kind`/`window` information already?

File: modules/gapi/src/backends/fluid/gfluidbackend.cpp [hunk 14]
```diff
@@ -1016,14 +1022,14 @@ namespace
                     if (d.shape == cv::GShape::GMAT)
                     {
                         auto port = g.metadata(in_edge).get<Input>().port;
-                        fu.line_consumption[port] = maxLineConsumption(fu.k, in_h, out_h, fu.k.m_lpi, port);
+                        fu.line_consumption[port] = maxLineConsumption(fu.k.m_kind, fu.window, in_h, out_h, fu.k.m_lpi, port);
```
Comment (dmatveev) [inline modules/gapi/src/backends/fluid/gfluidbackend.cpp]: oh, I see. Stuff partly goes from `fu.k` and partly from `fu` itself. OK

**Comment by dmatveev** [comment]
The last thing to address is to write a good MR description and explain what is changed and how. Thanks!

File: modules/gapi/src/backends/fluid/gfluidbackend.cpp (hunk 13 above)
Comment (akhakim) [reply modules/gapi/src/backends/fluid/gfluidbackend.cpp]: The fact is that now GFluidKernel doesn't contain a "window" field. According to Ruslan's suggestion, I moved the "window" field from GFluidKernel to a more specific and convenient for further use structure GFluidUnit. I think duplication is useless to us in this case. So, I was had to change signature of the borderSize() function.

**Pull request merged**